### ISSUE TITLE
Reduce bandwidth by doing eventual consistency only after a certain amount of time

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#125]
Joy — 12/31/2023 3:48 PM on Discord

traditionally the server maintains some bookkeeping per-entity, per-component (sometimes per field/property), i.e. it'll remember the last tick it sent an (entity, component) to a specific client, the last tick that client acknowledged, and the last tick that state changed on the server 
and that's what would be used to maintain eventual consistency

imagine your policy only chooses the entities to send but leaves which components to include up to the server
then, if sending a snapshot for tick N
if entity e's component T has tick changed > tick sent, include (e, T) in the snapshot and bump its tick sent to N
later
if this message is delivered (ACK), bump (e, T) tick acknowledged to N
if this message is lost (NACK), reset (e, T) tick sent to tick acknowledged
 
while not a literal binary diff of the component data, this is still a delta in a sense 
but yeah, in the absence of any user policy, the server could prioritize entities in descending order of "staleness" 

the ELI5 of this is "the server includes components that changed since its last send attempt"
the > is on "tick sent" and not "tick acknowledged" because we assume that the snapshot will be delivered
(the vast majority of packets do successfully arrive)
and that way, the server doesn't waste bandwidth redundantly spamming the same data
If the snapshot is lost, then it just resets its bookkeeping so that data once again falls under the send criteria

**Comment by cBournhonesque** [comment]
This is also causing some issues for me when writing a unit test for prediction:
- I create an Entity with component C = 1.0 on the server and replicate it at tick T
- on the client, I receive it at tick T, but then I also receive it at ticks T+1, T+2, etc. because the server keeps replicating its local version of component C until it receives an ack from the client.

- One possibility was to send the component updates for C since the most recent of either:
   - the last ack it received from the client
   - the last EntityAction (reliable) message it sent for the replication group
   
I think there was an issue with this second option, but I cannot remember what....

**Comment by cBournhonesque** [comment]
I implemented this, but we only keep the tick_send/tick_ack per entity instead of per (entity, component)

**Comment by cBournhonesque** [comment]
Actually keeping the `tick_send/tick_ack` per entity instead of per component is a bug!

Imagine you have:
- tick 0: spawn C1, C2. ack_tick = 0
- tick 2: update C1. send_tick = 2
- tick 3: update C2. send_tick = 3
- tick 6: NACK C1. send_tick = ack_tick = 0. Possibly re-send update for C1 -> send_tick = 6. Re-send update for C2 (useless!)
- tick 7: ACK C2. ack_tick = 3, send_tick = 0.

Then:
- at tick 7 it looks like we can guarantee that the remote state for tick 3 has the updates for C1 and C2, but actually it only has the update for C2!
- send_tick is still 0! (but it should be equal to 3 instead). (we could set send_tick to ack_tick if ack_tick is bigger..)

We could also have
Imagine you have:
- tick 0: spawn C1, C2. ack_tick = 0
- tick 2: update C1. send_tick = 2
- tick 3: update C2. send_tick = 3
- tick 6: ACK C2. ack_tick = 3, send_tick = 3.
- tick 7: NACK C1. send_tick = 3. We don't re-send an update for C1 since the update happened on tick 2 and the ack_tick is 3! The C1 update is lost.


By contrast, let's say we keep the send_tick/ack_tick per component:
Imagine you have:
- tick 0: spawn C1, C2. ack_tick = 0
- tick 2: update C1. send_tick_C1 = 2
- tick 3: update C2. send_tick_C2 = 3
- tick 6: NACK C1. send_tick_C1 = 0. Possibly re-send update for C1: send_tick_C1 = 6
- tick 7: ACK C2. ack_tick_C2 = 3, send_tick_C1 = 0.

Now we send updates correctly, but we don't have guarantees anymore that the entity is in a consistent compared to what it was on the server!


Basically this policy makes it possible to have entities in a not-consistent state. (i.e. the entity on the receiver has a state that was not equal to a past sender's state)

**Comment by cBournhonesque** [comment]
Maybe we can have another option:
- `send_tick`/`ack_tick` computed per entity
- send updates since `send_tick`. If there is at least 1 update, gather all updates since `ack_tick`. The issue is
- every time we send, update `send_tick`
- if NACK, send back `send_tick` back to `ack_tick`.

Imagine you have:
tick 0: spawn C1, C2. ack_tick = 0
tick 2: update C1. send_tick = 2
tick 3: update C2. send_tick = 3. We also send an update for C1 since we now gather all updates since `ack_tick`.
tick 6: NACK C1. send_tick = ack_tick = 0. If we re-send, we will send updates for C1 and C2.
tick 7: ACK C2. ack_tick = 3. Also set send_tick to 3 (send_tick is at least ack_tick)

Or:
tick 0: spawn C1, C2. ack_tick = 0
tick 2: update C1. send_tick = 2
tick 3: update C2. send_tick = 3. We also send an update for C1 since we now gather all updates since `ack_tick`.
tick 6: ACK C2. ack_tick = 3. Also set send_tick to 3 (send_tick is at least ack_tick). We know the entity is consistent and on tick 3.
tick 7: NACK C1. The send_tick for that message was lower than the ack_tick, so do nothing. (we know the entity is at a state that is more recent)
